_posts/2023/2023-01-22-only-use-locals-in-view-partials.md

```diff
@@ -0,0 +1,77 @@
+---
+title: "Only use locals in view partials"
+description: "It helps keep your state manageable"
+layout: article
+category: ruby
+image:
+  base: '2023/only-use-locals-in-view-partials'
+  alt: "Local building sign"
+  credit: "Priscilla Du Preez"
+  source: "https://unsplash.com/photos/acNPOikiDRw"
+
+---
+
+The Rails view layer is incredibly flexible, but can easily become tricky to manage even though it’s the main way our users and customers access our work.
+
+https://guides.rubyonrails.org/layouts_and_rendering.html#using-partials
+
+Many Ruby teams 
+
+## Instead of…
+
+…accessing instance variables inside partials:
+
+```ruby
+# calling the partial
+<%= render "user_email" %>
+
+# inside the _user_email.html.erb partial
+<%= @user.email %>
+```
+
+
+## Use…
+
+…local variables to pass in context:
+
+```ruby
+# calling the partial
+<%= render partial: "user_email", locals: { user: @user } %>
+
+# inside the _user_email.html.erb partial
+<%= user.email %>
+```
+
+
+## Why?
+
+This is primarily a good rule of thumb for maintenance and organisation in more complex projects. Rails explicitly allows a vast global scope in it's view environment, but many experienced developers (including me) have got themselves in trouble by reaching outside of a partial for a variable.
+
+Additionally, by not using instance variables or “helper-ed” methods (like `current_user` or Rails’s CurrentAttributes) within your partials you make them more reliably cache-able, because the values you’re passing into them will be all the context required to build a cache key. Without this it would be possible to accidentally serve a cached partial with incorrectly scoped data.
+
+For example:
+
+```ruby
+# calling the partial
+<%= render partial: 'things/thing', collection: @things, cached: true %>
+
+# things/_thing.html.erb partial
+<%= thing.name %>
+<%= link_to "Edit", edit_thing_path(thing) if @user.admin? %>
+```
+
+Multiple users will get the same partial pulled from the cache, either with the link to “Edit” or without, depending on which user rendered the page first. In addition you might not want the non-admin users to see the link, exposing a potential security hole.
```

Comment:
So the main problem here is line 60 and the reference to `@user` in the partial, right? I think that could be made clearer!